File: 2021/07/30/MaxProfit.java
/**
 * https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/
 * @author Andrew Jarombek
 * @since 7/30/2021
 */

public class MaxProfit {
    public static int maxProfit(int[] prices) {
        int totalProfit = 0;
        int start = Integer.MAX_VALUE;
        int last = Integer.MIN_VALUE;

        for (int price: prices) {
            if (price >= start && price >= last) {
                last = price;
            } else {
                if (last != Integer.MIN_VALUE && last - start > 0) {
                    totalProfit += last - start;
                }

                start = price;
                last = Integer.MIN_VALUE;
            }
        }

        if (last != Integer.MIN_VALUE && last - start > 0) {
            totalProfit += last - start;
        }

        return totalProfit;
    }

    public static void main(String[] args) {
        assert maxProfit(new int[] {1, 2, 3, 4, 5}) == 4;
        assert maxProfit(new int[] {7, 1, 5, 3, 6, 4}) == 7;
    }
}

File: 2021/07/30/max_profit.py
"""
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
Author: Andrew Jarombek
Date: 7/30/2021
"""

from typing import List


def max_profit(prices: List[int]) -> int:
    lowest = prices[0]
    highest = lowest
    lowest_after_highest = lowest

    for price in prices[1:]:
        if price < lowest and price < lowest_after_highest:
            lowest_after_highest = price
        elif price >= highest:
            if lowest > lowest_after_highest:
                lowest = lowest_after_highest
                lowest_after_highest = price

            highest = price
        elif price - lowest_after_highest > highest - lowest:
            lowest = lowest_after_highest
            highest = price
            lowest_after_highest = price

    return highest - lowest


if __name__ == '__main__':
    assert max_profit([7, 1, 5, 3, 6, 4]) == 5
